numcodecs/compat.py

```diff
@@ -27,80 +26,161 @@
+            # type information will not have been propagated via the old-style buffer
+            # interface, so we have to manually hack it back in after the fact
+            arr = arr.view(buf.typecode)
+
+    if dtype is not None:
+        # view as requested dtype
+        arr = arr.view(dtype)
+
+    return arr
+
+
+def ensure_contiguous_ndarray(buf, dtype=None):
+    """Convenience function to coerce `buf` to a numpy array, if it is not already a
+    numpy array. Also ensures that the returned value exports fully contiguous memory,
+    and supports the new-style buffer interface.
+
+    Parameters
+    ----------
+    buf : array-like or bytes-like
+        A numpy array or any object exporting a buffer interface.
+    dtype : dtype, optional
+        Request that the data be viewed as the given dtype.
+
+    Returns
+    -------
+    arr : ndarray
+        A numpy array, sharing memory with `buf`.
+
+    Notes
+    -----
+    This function will not create a copy under any circumstances, it is guaranteed to
+    return a view on memory exported by `buf`.
+
+    """
+
+    # ensure input is a numpy array
+    arr = ensure_ndarray(buf, dtype=dtype)
+
+    # check for datetime or timedelta ndarray, the buffer interface doesn't support those
+    if isinstance(buf, np.ndarray) and buf.dtype.kind in 'Mm':
+        arr = arr.view(np.int64)
+
+    # check for object arrays, these are just memory pointers, actual memory holding
+    # item data is scattered elsewhere
+    if arr.dtype == object:
+        raise TypeError('object arrays are not supported')
+
+    # check memory is contiguous, if so flatten
+    if arr.flags.c_contiguous or arr.flags.f_contiguous:
+        # can flatten without copy
+        arr = arr.reshape(-1, order='A')
 
-def ndarray_from_buffer(buf, dtype):
-    if isinstance(buf, np.ndarray):
-        arr = buf.reshape(-1, order='A').view(dtype)
     else:
-        arr = np.frombuffer(buf, dtype=dtype)
+        raise ValueError('an array with contiguous memory is required')
+
     return arr
 
 
-def ensure_text(l, encoding='utf-8'):
-    if isinstance(l, text_type):
-        return l
-    else:  # pragma: py3 no cover
-        return text_type(l, encoding=encoding)
+def ensure_bytes(buf):
+    """Obtain a bytes object from memory exposed by `buf`."""
+
+    if not isinstance(buf, binary_type):
+
+        # go via numpy, for convenience
+        arr = ensure_ndarray(buf)
 
+        # create bytes
+        buf = arr.tobytes(order='A')
 
-def handle_datetime(buf):
-    if hasattr(buf, 'dtype') and buf.dtype.kind in 'Mm':
-        return buf.view('i8')
     return buf
+
+
+def ndarray_copy(src, dst):
+    """Copy the contents of the array from `src` to `dst`."""
+
+    if dst is None:
+        # no-op
+        return src
+
+    # ensure ndarrays
+    src = ensure_ndarray(src)
+    dst = ensure_ndarray(dst)
+
+    # ensure same data type
+    if dst.dtype != object:
```

Comment:
I can't remember if having ``dst`` with ``object`` type is specifically covered. Actually it's an area where coverage could be improved, what happens when different shapes and types of array are provided as the ``out`` argument to ``codec.decode()``. We could remove the check, in which case an error would get raised on any attempt to view as object dtype. I think I left the check because then it would be possible to copy data into an object array, although I'm not entirely sure that's something we want to support. I believe ``np.copyto()`` can handle object arrays in a sensible way.